### PR TITLE
Virtual screening nab

### PR DESCRIPTION
**Pull request by flamingofugang**
The commits aim to add NIM Agent Blueprint end-2-end example running on AWS HealthOmics.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: workflows/virtual-screening-nab/definition/main.nf [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        params.diffdock_model,
+        RunMolMIMGenerate.out.sdfs,
+        RunAlphaFold2.out.pdfs,
+        num_poses
+    )
+}
+
+process RunMolMIMGenerate {
+    label 'molmim'
+    cpus 8
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/definition/main.nf:51]: we should confirm with NVIDIA if 4/16 or 8/32 is better

File: workflows/virtual-screening-nab/definition/main.nf [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    script:
+    """
+    set -euxo pipefail
+
+    cp /opt/*.py .
+    export PYTHONPATH="${PYTHONPATH}:/opt/:./"
+    export NIM_CACHE_PATH=${model_parameters}
+    for seq in `cat ${sequences}`
+    do
+        python ${executable} \$seq
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/definition/main.nf:108]: same question as above wrt executables/setups

File: workflows/virtual-screening-nab/definition/main.nf [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    script:
+    """
+    set -ex
+
+    cp /opt/*.py .
+    export NIM_CACHE_PATH=${model_parameters}
+    export PYTHONPATH="${PYTHONPATH}:/opt/./"
+    /usr/local/lib/python3.10/dist-packages/nimlib/triton_start.sh -m ${model_parameters}/bionemo-diffdock_v1.2.0 -p 8080 -l 0 &
+    Triton_PID=\$!
+    sleep 300
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/definition/main.nf:140]: can probably shrink this down to 120 and then poll

File: workflows/virtual-screening-nab/definition/main.nf [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    for pro in ${protein_file}
+    do
+        for lig in ${ligand_file}
+        do
+            /usr/bin/python3 ${executable} \$pro \$lig ${num_poses}
+            sleep 10
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/definition/main.nf:146]: sleep 10 feels long. maybe sleep 1?

File: workflows/virtual-screening-nab/definition/params.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+    "smi_input_path":"s3://<mybucket>/molmim/input/",
+    "fasta_path":"s3://<mybucket>/alphafold/input/"
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/definition/params.json:3]: we'll want to add additional params here (eg number of predictions)

File: workflows/virtual-screening-nab/definition/main.nf [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        smi_input_path = params.smi_input_path
+    }
+
+    if (params.fasta_path[-1] == "/") {
+        fasta_path = params.fasta_path + "*"
+    } else {
+        fasta_path = params.fasta_path
+    }
+
+    smiles = Channel.fromPath(smi_input_path)
```
[the remark this answers is not in the document]
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf:18]: The smi input could be a list of SMILES strings. It is not clear about the size limit though.

File: workflows/virtual-screening-nab/definition/main.nf [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        params.alphafold2_script,
+        sequences
+    )
+
+    num_poses = 3
+    RunDiffdock(
+        params.diffdock_script,
+        params.diffdock_model,
+        RunMolMIMGenerate.out.sdfs,
+        RunAlphaFold2.out.pdfs,
```
[the remark this answers is not in the document]
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf:44]: pdbs

File: workflows/virtual-screening-nab/definition/main.nf [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    for s in `cat ${smiles}`
+    do
+        echo \$s
+
+        python ${molmim_script} ${smiles} .
+        sleep 1
+    done
+
+    sleep 60
```
[the remark this answers is not in the document]
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf:78]: we can delete it

File: workflows/virtual-screening-nab/definition/main.nf (hunk 1 above)
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf:51]: implemented a retry to go with 4/16 first then 8/32

File: workflows/virtual-screening-nab/definition/main.nf (hunk 3 above)
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf:140]: changed to 120 seconds, but curl localhost:8080 does not work for tritonserver endpoint

File: workflows/virtual-screening-nab/README.md [hunk 9]
```diff
@@ -77,7 +77,7 @@ ROLEARN=
 OUTPUTLOC=s3://<mybucket>/nab/output/
 PARAMS=./params.json
 
-aws omics start-run --workflow-id $WFID --role-arn $ROLEARN --output-uri $OUTPUTLOC --storage-type DYNAMIC --parameters file://definition/params.json --name nim-nab
+aws omics start-run --workflow-id $WFID --role-arn $ROLEARN --output-uri $OUTPUTLOC --storage-type STATIC --parameters file://definition/params.json --name nim-nab
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/README.md:80]: if STATIC, will need a storage capacity as well (4800)

File: workflows/virtual-screening-nab/README.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
+# Generative Virtual Screening NIM Agent Blueprint on AWS HealthOmics
```
Comment (ajfriedman18) [inline workflows/virtual-screening-nab/README.md:1]: NVIDIA Generative Virtual Screening Blueprint on AWS HealthOmics

File: workflows/virtual-screening-blueprint/definition/config.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+name: NIMAgentBlueprint
```
Comment (ajfriedman18) [inline workflows/virtual-screening-blueprint/definition/config.yaml:1]: NVIDIAVirtualScreeningBlueprint

File: workflows/virtual-screening-blueprint/definition/config.yaml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    description: "Input SMILES."
+    optional: false
+  fasta_path: 
+    description: "Input file in FASTA format."
+    optional: false
+  num_molecules:
+    description: "Number of molecules to generate by MolMIM."
+    optional: false
+  num_poses:
+    description: "Number of docking poses."
```
Comment (ajfriedman18) [inline workflows/virtual-screening-blueprint/definition/config.yaml:16]: add "generated by DiffDock"

File: workflows/virtual-screening-blueprint/definition/main.nf [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        RunMolMIMGenerate.out.sdfs,
+        RunAlphaFold2.out.pdbs,
+        params.num_poses
+    )
+}
+
+process RunAlphaFold2 {
+    label 'alphafold2'
+    errorStrategy 'retry'
+    cpus { 2 * Math.pow(2, task.attempt+2) }
```
Comment (ajfriedman18) [inline workflows/virtual-screening-blueprint/definition/main.nf:53]: this looks off/too high. I think this is saying the minimum resources required here are 16 vCPUs and 64GB memory on (g5.4xlarge).

Comment (ajfriedman18) [reply workflows/virtual-screening-blueprint/definition/main.nf:53]: for both cpus and memory this should remove the `+2`

File: workflows/virtual-screening-blueprint/definition/main.nf [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    memory { 8.GB * Math.pow(2, task.attempt+2) }
+    accelerator 1, type: 'nvidia-tesla-a10g'
+    maxRetries 1
```
Comment (ajfriedman18) [inline workflows/virtual-screening-blueprint/definition/main.nf:56]: up this to 2...longer term this should be a param in `nextflow.config`

**Review by ajfriedman18 — CHANGES_REQUESTED** [review]

File: workflows/virtual-screening-nab/definition/main.nf [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    smiles = Channel.fromPath(smi_input_path)
+        .collectFile(name: 'smi_input.json')
+
+    RunMolMIMGenerate(
+        params.molmim_script,
+        params.molmim_model,
+        smiles
+    )
+
+    sequences = Channel.fromPath(fasta_path)
```
[the remark this answers is not in the document]
Comment (flamingofugang) [reply workflows/virtual-screening-nab/definition/main.nf]: I put Alphafold process earlier than MolMIM process in the workflow, but it looks like MolMIM still run first...

File: workflows/virtual-screening-blueprint/README.md [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+* `s3://{mybucket}/alphafold2/`
+* `s3://{mybucket}/alphafold2/input/`
+* `s3://{mybucket}/diffdock/`
+* `s3://{mybucket}/nim/scripts/`
+* `s3://{mybucket}/nim/output/`
+
+First upload python scripts (in `scripts` folder) to run inside NIM containers, including `molmim_generate.py`, `alphafold2_predict.py`, and `run_diffdock.py` to S3 path `s3://{mybucket}/nim/scripts/`. 
+
+Secondly, upload sample SMILES and FASTA files in the `input` folder to s3:
+* `s3://{mybucket}/molmim/input/inputsmiles1.json`
```
[the remark this answers is not in the document]
Comment (flamingofugang) [reply workflows/virtual-screening-blueprint/README.md:38]: I have changed it to smi file instead of json file

File: workflows/virtual-screening-blueprint/definition/main.nf (hunk 13 above)
Comment (flamingofugang) [reply workflows/virtual-screening-blueprint/definition/main.nf:53]: The recommended mem to run alphafold2 is 64GB, I think I got OOR using 16GB, let me try again for 32GB mem.

Comment (ajfriedman18) [reply workflows/virtual-screening-blueprint/definition/main.nf:53]: got it - if that's what they recommend can stay with it
